### PR TITLE
[IMP] base,web: use SHA2 over SHA1

### PR DESCRIPTION
**Pull request by Julien00859**
SHA-1 is a cryptographic hash function that have weaknesses known since
2005, it has been deprecated by the NIST [1] about 10 years ago in 2011
and Google [2] have been able to perform a collision attack in 2017.

We use SHA-1 to compute the checksum of every attachment and reuse that
checksum in order to store the file on disk. The purpose is only store
the file once and possibly link that file in many attachments.

We use SHA-1 in order to generate unique URL for resources that can be
cached by the browser: assets bundle, translations, qweb templates and
qweb images.

Although practical attacks still requires quite a lot of computational
resources, it is time to upgrade SHA-1 to SHA-2.

To prevent attacks on the filestore, when uploading a file that has
the same SHA-1 as a stored file, we perform a SHA-2 computation
on both files to verify their signature. If their SHA-2 signature is 
different, an error is raised.

We have selected the SHA-512/256 variant of the SHA-2 algorithm as
replacement for SHA-1 for the following reasons:

* On 64 bits platform, SHA-512 is the fastest SHA-2 variant, it is only
  ~1.5x slower than SHA-1. [3]
* Keeping only the 256 foremost bits protects against both collision
  attacks and length extension attacks.
* The hexadecimal digest is only 24 chars longer than SHA-1 which is
  nice to have somewhat short URLs.

We have not used SHA-3 because:

* At the moment of writing, it is too slow (~3x slower than SHA-1) [3]
* It is not guaranteed to be available with the Python 3.5 `hashlib`
  module.
* One of the author of SHA-3 is Belgian.

[1] https://csrc.nist.gov/projects/hash-functions/nist-policy-on-hash-functions
[2] https://shattered.io/
[3] http://bench.cr.yp.to/results-hash.html
[4] http://www.commitstrip.com/en/2017/02/27/the-sha-1-alternative/

task-31589

### CODE REVIEW

**Comment by Yenthe666** [comment]
> One of the author of SHA-3 is Belgian.
you guys 😆

**Comment by Julien00859** [comment]
+ removed index on the checksum field

**Comment by VincentSchippefilt** [comment]
note that I verified only /web/ and it tested good, cache is still working as expected

**Comment by mreficent** [comment]
why not point to saas12.5?

**Comment by Julien00859** [comment]
> why not point to saas12.5?

We are speaking about migrating filestores to prevent not common SHA-1 collisions, I ensure you guys rather we test it extensively before it gets merged in a stable version and I'm also about sure there will be some welcome tuning :wink:

**Comment by Julien00859** [comment]
* inverted the two `self.env['ir.attachment']._file_gc()` and `self.env['ir.filestore_sha1_migration']._vacuum(100)` so the vacuum of sha1 marks files for garbage collection the next day

**Comment by Julien00859** [comment]
* replaced the filestore SHA-2 migration by raising an error when two files collides

File: odoo/addons/base/models/ir_attachment.py [hunk 1]
```diff
@@ -88,6 +88,15 @@ def _get_path(self, bin_data, sha):
         dirname = os.path.dirname(full_path)
         if not os.path.isdir(dirname):
             os.makedirs(dirname)
+        # prevent sha-1 collision
+        if os.path.isfile(full_path):
+            stored_sha2 = self._compute_file_checksum(full_path, hashlib.sha512)
+            new_sha2 = hashlib.sha512(bin_data or b'').hexdigest()
+            if stored_sha2 != new_sha2:
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_attachment.py]: Why not comparing both contents byte-by-byte?  Hashing them is just much slower...
I don't see any added value in having their SHA-2.

Comment (Julien00859) [reply odoo/addons/base/models/ir_attachment.py]: done

**Comment by Julien00859** [comment]
@robodoo rebase-ff

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

File: odoo/addons/base/models/ir_attachment.py (hunk 1 above)
Comment (Julien00859) [reply odoo/addons/base/models/ir_attachment.py]: @rco-odoo see, there was SHA-512 in `_same_content` at one point! :D

**Pull request closed without merging**